### PR TITLE
Added new API for low precision fp8 attention using FA3

### PR DESCRIPTION
**Pull request by howardzhang-cv**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.13.0) (oldest at bottom):
* #3947
* #3960
* #3930
* #3929
* #3865
* __->__ #3857
* #3959

 ---                                                                                                                                                                                                                                               
## Summary                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          
  - Added RoPE fusion compile path for FA3 FP8 low-precision attention (fuse_rope=True)                                                                                                                                                             
  - New elementary block: fp8_fa3_rope_sdpa — fused RoPE + FP8 quantization + low-precision SDPA
  - New Triton kernel for fused RoPE + QKV quantization with layout transpose ([B,S,H,D] → [B,H,S,D])
  - RoPE fusion method: Custom Inductor backend that traces the FX graph, detects RoPE + SDPA patterns (NeoX half-split and FLUX interleaved formats), and replaces them with fp8_fa3_rope_sdpa custom ops. Falls back to fp8_fa3_sdpa for SDPA
  nodes without RoPE.
  - Causal mask detection: Pre-flight forward pass identifies HuggingFace-style materialized causal masks so the fusion pass can strip them and use is_causal=True instead.
  - Added compiled model wrapper (_FP8FlashAttentionCompiledWrapper) with @torch._dynamo.disable to prevent re-tracing.
  - Added RoPE SDPA numerical accuracy tests and fuse_rope parametrization on model-level tests.

### New Files
  - shared_utils/fusion_utils.py: Shared FX graph fusion pass — RoPE pattern detection, SDPA detection, transpose unwrapping, parameterized graph surgery
  - shared_utils/custom_ops.py: Factory functions to register backend-specific custom ops with register_fake, and helpers to build fusion passes and compile functions
  - fp8_fa3/fusion_pass.py: FA3-specific custom op registration, rope_sdpa_fusion_pass, and compile_with_fp8_fusion entry point
  - quantization/triton_rope_qkv_quantization.py: Fused RoPE + QKV FP8 quantization Triton kernel

### Modified Files
  - shared_utils/attention.py: Added _fp8_rope_sdpa shared implementation
  - shared_utils/wrapper.py: Added _FP8FlashAttentionCompiledWrapper
  - shared_utils/setup.py: Added compile path routing via compile_fn parameter, moved detect_causal_mask to fusion_utils.py
  - quantization/quantization.py: Added _fp8_rope_sdpa_quantize
  - fp8_fa3/attention.py: Added fp8_fa3_rope_sdpa elementary block
  - fp8_fa3/setup.py: Passes compile_with_fp8_fusion as compile_fn
  - test_fp8_attention.py: Added TestFP8RopeSDPANumericalAccuracy, fuse_rope parametrization on model test

## Test Plan
`python -m pytest test/prototype/attention/test_fp8_attention.py -v`

## Example Usage
```python
  from torchao.prototype.attention import (
      AttentionBackend,
      LowPrecisionAttentionConfig,
      apply_low_precision_attention,
  )

  model = MyModel()

  # Enable low precision attention with config
  config = LowPrecisionAttentionConfig(
      backend=AttentionBackend.FP8_FA3,
  )
  model = apply_low_precision_attention(model, config)
  # Run torch.compile automatically applies rope fusion
  model = torch.compile(model)

  # Flash activation is handled internally by the wrapper
  output = model(inputs)

```
  ---
## Results
#### Single-Layer Results
Results directly comparing FA3 SDPA versus FA3 fp8 SDPA (including quantization time):
<img width="645" height="373" alt="image" src="https://github.com/user-attachments/assets/f64cba6d-4ac9-41b7-b0f7-bf93c67f6c13" />
#### Llama3 Model Results
Results comparing Llama3 model with FA3 SDPA versus Llama3 using the FA3 fp8 wrapper. Uses RoPE fusion.
Perplexity: 6.19 -> 6.24
<img width="634" height="285" alt="image" src="https://github.com/user-attachments/assets/7e454bdc-b0d2-43e6-8293-32012abde5f8" />

### CODE REVIEW

**Comment by howardzhang-cv** [comment]
Assuming this is gonna take a bit of back and forth to land to get it just right, since this is a user-facing change. @drisspg please take a look and let me know if the API looks okay

File: torchao/prototype/attention/fp8_fa3/wrappers.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    config: Optional[LowPrecisionAttentionConfig] = None,
+) -> nn.Module:
+    """Wrap model so its forward pass uses FP8 FA3 attention."""
+    if config is None:
+        config = LowPrecisionAttentionConfig()
+
+    original_forward = model.forward
+
+    def wrapped_forward(*args, **kwargs):
+        with _fp8_fa3_attention_context(config):
```
Comment (jerryzh168) [inline torchao/prototype/attention/fp8_fa3/wrappers.py]: not sure if we talked about this before, what's the thoughts on this v.s. following other torchao APIs, e.g.
```
q = Float8Tensor.from_hp(q)
k = Float8Tensor.from_hp(k)
v = Float8Tensor.from_hp(v)
# dispatch to `_fp8_fa3_sdpa` in Float8Tensor implementation
F.scaled_dot_product_attention(q, k, v, is_causal=True)
```
?

Comment (howardzhang-cv) [reply torchao/prototype/attention/fp8_fa3/wrappers.py]: Yeah, ideally we follow other torchao APIs but it's a bit difficult for attention. This first backend/recipe works well because it is a simple replacement of F.scaled_dot_product_attention, but as we add other features such as RoPE fusion or RoPE + Hadamard it becomes much more difficult, since it becomes model specific (i.e. if there is RoPE followed by SDPA, replace with RoPE + quantization fused kernel into fp8 SDPA). To future-proof other attention backends and recipes, I think it's better to have it as its own separate APIs. What are your thoughts?

Comment (jerryzh168) [reply torchao/prototype/attention/fp8_fa3/wrappers.py]: I thought the fused kernel replacement should happen in inductor? or is there a reason why we have to hand replace these in eager mode?

Comment (howardzhang-cv) [reply torchao/prototype/attention/fp8_fa3/wrappers.py]: Inductor unfortunately does not fuse RoPE with the quantization kernel

Comment (jerryzh168) [reply torchao/prototype/attention/fp8_fa3/wrappers.py]: oh I mean is this something that's not possible to do in inductor or just something that does not exist right now

Comment (howardzhang-cv) [reply torchao/prototype/attention/fp8_fa3/wrappers.py]: Not too sure, I feel like it should be possible in inductor? But it definitely is something that does not exist right now. Could look into it for the future, but for now, I think it would be simpler just to move this into prototype and have it be available through that. Will work on moving this over.

File: torchao/prototype/attention/fp8_fa3/quantization.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    if k.shape != v.shape:
+        raise ValueError(f"K and V shape mismatch: {k.shape} vs {v.shape}")
+    if q.shape[0] != k.shape[0]:
+        raise ValueError(f"Batch size mismatch: {q.shape[0]} vs {k.shape[0]}")
+    if q.shape[1] != k.shape[1]:
+        raise ValueError(f"Head count mismatch: {q.shape[1]} vs {k.shape[1]}")
+    if q.shape[3] != k.shape[3]:
+        raise ValueError(f"Head dim mismatch: {q.shape[3]} vs {k.shape[3]}")
+
+    if torch.compiler.is_compiling():
```
Comment (vkuzo) [inline torchao/prototype/attention/fp8_fa3/quantization.py]: is there a test comparing numerics between these two paths

File: torchao/prototype/attention/__init__.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+quantization strategies and different PyTorch core attention backends.
+
+Note: Low-precision attention only supports inference (forward pass).
+Backward pass is not supported by the underlying backends.
+
+Example::
+
+    from torchao.prototype.attention import apply_low_precision_attention
+
+    model = MyTransformer()
```
Comment (vkuzo) [inline torchao/prototype/attention/__init__.py]: we should specify what kind of syntax is automatically converted to low precision here.  F.SDPA? something else?

File: torchao/prototype/attention/api.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        )
+
+    from torch._inductor.compile_fx import compile_fx
+
+    from torchao.prototype.attention.fp8_fa3.fusion_pass import (
+        rope_sdpa_fusion_pass,
+    )
+
+    def fp8_attention_backend(gm, example_inputs):
+        """Custom Inductor backend that applies the RoPE + FP8 fusion pass."""
```
Comment (vkuzo) [inline torchao/prototype/attention/api.py]: this should specify what kind of user syntax counts as "attention"

File: torchao/prototype/attention/config.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+
+class AttentionBackend(str, Enum):
+    """
+    Backend kernel for computing attention.
+
+    Different backends have different hardware requirements and capabilities.
+    """
+
+    FP8_FA3 = "fa3"
```
Comment (vkuzo) [inline torchao/prototype/attention/config.py]: make the string match the enum value name

File: torchao/prototype/attention/api.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        # cached and FA3 activation is a harmless no-op for the fused
+        # custom ops (they select the backend internally).
+        activate_flash_attention_impl("FA3")
+        try:
+            return compiled_mod(*args, **kwargs)
+        finally:
+            restore_flash_attention_impl()
+
+
+def apply_low_precision_attention(
```
Comment (vkuzo) [inline torchao/prototype/attention/api.py:60]: I think this should be explicit that parts of torch.compile is used to do the logic swap

File: torchao/prototype/attention/fp8_fa3/quantization.py (hunk 2 above)
Comment (howardzhang-cv) [reply torchao/prototype/attention/fp8_fa3/quantization.py]: The numerics are exactly the same between the two paths. I tested the runtime and it seems like the triton implementation was much faster (~100 ms difference on llama3 model with 124k sequence length)

File: torchao/prototype/attention/fp8_fa3/wrappers.py (hunk 1 above)
Comment (howardzhang-cv) [reply torchao/prototype/attention/fp8_fa3/wrappers.py]: @jerryzh168 I've gotten the rope fusion to work by using the inductor path. I've moved everything to a prototype folder. Due to the way attention works, I think it's better if it's own API, as it doesn't fit cleanly with others for now (especially with rope fusion and hadamard and stuff requiring inductor path). What do you think?

File: torchao/prototype/attention/__init__.py (hunk 3 above)
Comment (howardzhang-cv) [reply torchao/prototype/attention/__init__.py]: Added

File: torchao/prototype/attention/api.py (hunk 4 above)
Comment (howardzhang-cv) [reply torchao/prototype/attention/api.py]: Added

File: torchao/prototype/attention/config.py (hunk 5 above)
Comment (howardzhang-cv) [reply torchao/prototype/attention/config.py]: Fixed

File: torchao/prototype/attention/api.py (hunk 6 above)
Comment (howardzhang-cv) [reply torchao/prototype/attention/api.py:60]: Added, additionally added a warning as well to be even more explicit
